infra/templates/gitlab-bootstrap-iac-cicd/azure.tf

```diff
@@ -0,0 +1,62 @@
+resource "random_string" "rand" {
+  length  = 4
+  special = false
+  number  = false
+  upper   = false
+}
+
+resource "azurerm_resource_group" "ci" {
+  name     = format("rg-%s-ci", var.prefix)
+  location = var.location
+}
+
+resource "azurerm_storage_account" "ci" {
+  name                = format("backendstate%s", random_string.rand.result)
+  resource_group_name = azurerm_resource_group.ci.name
+  location            = azurerm_resource_group.ci.location
+
+  min_tls_version          = "TLS1_2"
+  account_tier             = "Standard"
+  account_replication_type = "LRS"
+}
+
+resource "azurerm_storage_container" "tfstate" {
+  name                  = "tfstate-terraform-bootstrap"
+  storage_account_name  = azurerm_storage_account.ci.name
+  container_access_type = "private"
+}
+
+data "azurerm_client_config" "current" {}
+
```

Comment:
This file is stamping out 3 environment modules. Did you face any challenges trying to collapse this into one?

Reply:
Nope. We could have also made the module consume a list of `environments` (i.e., `['dev', 'integration', 'prod']`) and also make all the other parameters a list, but it has some interesting behaviors that we want to avoid:

For example, in the current implementation we can simply add a new environment by copy + paste of the module block. If the the module instead took a list of environments, then the caller must be sure to add new environments only to the **end**:

i.e.,  `['dev', 'integration', 'prod', 'foo-env']` and not  `['dev', 'integration', 'foo-env', 'prod']` so that the apply operation is *create only*. The first example would achieve that, whereas the second example would teardown `prod`, before creating `foo-env` and `prod` resources.

Simply put, the interface we chose makes it much more difficult to accidentally delete resource groups that contain production deployments :)